File: web/src/api/free-discovery.ts
import { req } from './_core'

// free-discovery.ts — 免费资源自动发现 Admin API 封装 (2026-09-09)。
//
// 后端: admin/free_discovery.go + domains/freediscovery
// 文档: docs/freediscovery-configuration.md (含 §4 契约: SSRF / 禁用模板 / 任务状态机).
// 请求/响应字段全部 snake_case, 由后端 struct json tag 决定.
//
// 流程: 模板(含内置预设一键创建) → 触发扫描(/scan, 失败也返回 200+任务体)
//       → 审查发现结果 → 批量导入 free_resource_catalog (冲突策略可选)。
//
// 错误约定 (admin/free_discovery.go fdStatusFor):
//   - 400  校验/SSRF 阻断 (loopback / private / userinfo / fragment / 控制字符 /
//         scheme 不对 / models_endpoint 含 scheme·host·userinfo)
//   - 404  ErrTemplateNotFound / ErrImportTaskNotFound (含跨租户)
//   - 409  ErrTemplateDisabled / ErrImportTaskNotReady / ErrTaskStateConflict
//   - 500  兜底; scan 在 task 已 failed 时仍 200+任务体

export interface FreeDiscoveryTemplate {
  id: number
  tenant_id: string
  provider_code: string
  display_name: string
  base_url: string
  api_type: string // openai-completions | google-generative-ai | anthropic
  // api_key_env 为 "$VAR" 形式的环境变量引用; 空字符串 = keyless
  api_key_env: string
  models_endpoint: string
  quota_endpoint: string
  tos_url: string
  tos_verdict: string // ok | caution | ambiguous | avoid | unknown
  tos_notes: string
  enabled: boolean
  created_by: string
  created_at: string
  updated_at: string
}

export interface FreeDiscoveryTemplateCreate {
  provider_code: string
  display_name: string
  base_url: string
  api_type?: string
  api_key_env?: string
  /** 明文 Key (可选, 需后端已配 keyring; 无 keyring 时 fail closed) */
  api_key?: string
  models_endpoint?: string
  quota_endpoint?: string
  tos_url?: string
  tos_verdict?: string
  tos_notes?: string
  enabled?: boolean
}

export interface FreeDiscoveryTemplateUpdate {
  display_name?: string
  base_url?: string
  api_type?: string
  api_key_env?: string
  api_key?: string
  models_endpoint?: string
  quota_endpoint?: string
  tos_url?: string
  tos_verdict?: string
  tos_notes?: string
  enabled?: boolean
}

export interface FreeDiscoveryPreset {
  provider_code: string
  display_name: string
  base_url: string
  api_key_env: string
  api_type: string
  tos_verdict: string
  tos_notes: string
}

export interface FreeDiscoveryTask {
  id: number
  tenant_id: string
  template_id: number | null
  provider_code: string
  status: 'pending' | 'running' | 'success' | 'failed'
  trigger_type: 'manual' | 'scheduled' | 'webhook'
  triggered_by: string
  started_at: string | null
  completed_at: string | null
  error_message: string
  models_found: number
  models_imported: number
  created_at: string
  updated_at: string
}

export interface FreeDiscoveryResult {
  id: number
  task_id: number
  tenant_id: string
  provider_code: string
  model_id: string
  display_name: string
  context_window: number
  max_tokens: number
  free_type: string
  monthly_tokens: number
  daily_tokens: number
  pool_key: string
  tos_verdict: string
  tos_notes: string
  import_status: 'pending' | 'imported' | 'skipped' | 'conflict'
  imported_at: string | null
  created_at: string
}

export interface FreeDiscoveryImportSummary {
  imported: number
  skipped: number
  conflicted: number
  failed: number
}

export type FreeDiscoveryConflictPolicy = 'skip' | 'overwrite' | 'merge'

// ── 模板 ────────────────────────────────────────────────────────────────

export function listFreeDiscoveryTemplates(enabledOnly?: boolean): Promise<{ templates: FreeDiscoveryTemplate[] }> {
  const q = enabledOnly ? '?enabled=true' : ''
  return req('GET', `/api/free-discovery/templates${q}`)
}

/**
 * 创建模板.
 *
 * 400 校验失败:
 *   - provider_code 不符合 [a-z0-9-] 或长度 > 64
 *   - base_url 命中 SSRF (loopback / private / link-local / multicast / metadata /
 *     userinfo / fragment / 控制字符 / 非 http·https)
 *   - models_endpoint 含 scheme / host / userinfo (必须 `/` 开头的相对路径)
 *   - api_type / tos_verdict 非法
 */
export function createFreeDiscoveryTemplate(body: FreeDiscoveryTemplateCreate): Promise<FreeDiscoveryTemplate> {
  return req('POST', '/api/free-discovery/templates', body)
}

export function updateFreeDiscoveryTemplate(id: number, body: FreeDiscoveryTemplateUpdate): Promise<FreeDiscoveryTemplate> {
  return req('PATCH', `/api/free-discovery/templates/${id}`, body)
}

export function deleteFreeDiscoveryTemplate(id: number): Promise<{ deleted: boolean; id: number }> {
  return req('DELETE', `/api/free-discovery/templates/${id}`)
}

/**
 * 内置预设 (Groq / OpenRouter / Google AI Studio / SiliconFlow / Zhipu).
 *
 * 注意 google-ai-studio 预设的 api_type 固定为 `google-generative-ai`, 与其余
 * 4 家的 `openai-completions` 不同; UI 按 api_type 渲染扫描适配器.
 */
export function getFreeDiscoveryPresets(): Promise<{ presets: FreeDiscoveryPreset[] }> {
  return req('GET', '/api/free-discovery/templates/presets')
}

/**
 * 整体直传 Orbi pi-providers JSON: `{"providers": {"groq": {...}}}`.
 *
 * 返回 `status: ok | partial | failed` 表示整批导入结果: failed=0 时 ok,
 * 存在失败但成功 > 0 时 partial, 全失败时 failed. 单条失败不阻断其余.
 * 入参大小上限 1 MiB (admin handler enforce), 含 SSRF 校验.
 */
export function importFreeDiscoveryOrbiTemplate(
  file: { providers: Record<string, { baseUrl: string; api: string; apiKey: string }> },
): Promise<{ created: number; failed: number; errors: string[]; status: 'ok' | 'partial' | 'failed' }> {
  return req('POST', '/api/free-discovery/templates/import-orbi', file)
}

// ── 任务与结果 ──────────────────────────────────────────────────────────

/**
 * 触发扫描 (60s 独立上下文).
 *
 * 200 + 任务体: 任务已落库 (success / running / failed 均返回 200, UI 据 status 渲染).
 * 400 template_id 缺失; 404 ErrTemplateNotFound; 409 ErrTemplateDisabled.
 * 注: 上游失败时任务 status=failed 但 HTTP 仍 200, 错误见 task.error_message.
 */
export function scanFreeDiscovery(templateId: number): Promise<FreeDiscoveryTask> {
  return req('POST', '/api/free-discovery/scan', { template_id: templateId })
}

export function listFreeDiscoveryTasks(limit = 50): Promise<{ tasks: FreeDiscoveryTask[] }> {
  return req('GET', `/api/free-discovery/tasks?limit=${limit}`)
}

export function getFreeDiscoveryTask(id: number): Promise<FreeDiscoveryTask> {
  return req('GET', `/api/free-discovery/tasks/${id}`)
}

export function listFreeDiscoveryTaskResults(
  taskId: number,
  status: 'pending' | 'all' = 'pending',
): Promise<{ results: FreeDiscoveryResult[] }> {
  return req('GET', `/api/free-discovery/tasks/${taskId}/results?status=${status}`)
}

// ── 批量导入 ────────────────────────────────────────────────────────────

/**
 * 把任务下 pending 结果批量导入 free_resource_catalog.
 *
 * 200 + ImportSummary: 全部成功或 skip 跳过;
 * 404 ErrImportTaskNotFound (含跨租户);
 * 409 ErrImportTaskNotReady (task.status ≠ success) / ErrTaskStateConflict (并发覆盖).
 *
 * conflict_policy:
 *   - skip      (默认) 保留已有 catalog, 结果 mark skipped
 *   - overwrite 用发现结果覆盖 display/free_type/配额/ToS, 并重新启用
 *   - merge     仅补充 NULL/0/unknown 字段
 *
 * 单事务, 任一错误整体回滚; 重复扫描不重置已审查状态 (ON CONFLICT 保留
 * import_status / imported_at). catalog 写入 CAS: 0 行视为已被并发处理,
 * 计入 conflicted (不再静默 imported).
 */
export function importFreeDiscoveryResults(body: {
  task_id: number
  // 省略 result_ids = 导入该任务全部 pending
  result_ids?: number[]
  conflict_policy?: FreeDiscoveryConflictPolicy
}): Promise<FreeDiscoveryImportSummary> {
  return req('POST', '/api/free-discovery/import', body)
}

File: admin/free_discovery.go
package admin

// 免费资源自动发现 Admin API (/api/free-discovery/*).
//
// 数据模型: sql/migrations/084-freediscovery-schema.sql
// 领域包:   domains/freediscovery (模板 CRUD / 发现引擎 / 批量导入)
//
// 路由在 registerRoutes 中挂载; 服务依赖经 SetFreeDiscovery 在 main.go
// 启动时注入 (dbConn.Stdlib() 桥接 + credential keyring).

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/kaixuan/llm-gateway-go/domains/freediscovery"
	"github.com/kaixuan/llm-gateway-go/secret"
)

// freeDiscoveryDeps 聚合三个领域服务, 共享同一个 stdlib DB 桥接.
type freeDiscoveryDeps struct {
	templates *freediscovery.TemplateManager
	engine    *freediscovery.DiscoveryEngine
	importer  *freediscovery.ImportService
}

// SetFreeDiscovery 注入免费资源发现服务. 启动时调用一次;
// stdlibDB 为 nil (no-DB 模式) 时跳过, 相关路由在请求时返回 503.
func (h *Handler) SetFreeDiscovery(stdlibDB *sql.DB, kr *secret.Keyring) {
	if stdlibDB == nil {
		return
	}
	tm := freediscovery.NewTemplateManager(stdlibDB, kr)
	h.freeDiscovery = &freeDiscoveryDeps{
		templates: tm,
		engine:    freediscovery.NewDiscoveryEngine(stdlibDB, tm),
		importer:  freediscovery.NewImportService(stdlibDB),
	}
}

// fdDeps 取依赖; 未注入时写 503 并返回 nil.
func (h *Handler) fdDeps(w http.ResponseWriter) *freeDiscoveryDeps {
	if h.freeDiscovery == nil {
		writeError(w, http.StatusServiceUnavailable, "free-discovery is not available (database disabled)")
		return nil
	}
	return h.freeDiscovery
}

func (h *Handler) fdTenant(r *http.Request) string {
	// EffectiveTenantID: tenant_admin → 自有租户; super_admin/legacy → "default".
	return EffectiveTenantID(r)
}

// ── 模板管理 ────────────────────────────────────────────────────────────

// handleFreeDiscoveryTemplates GET(列表) / POST(创建).
func (h *Handler) handleFreeDiscoveryTemplates(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	switch r.Method {
	case http.MethodGet:
		enabledOnly := r.URL.Query().Get("enabled") == "true"
		tpls, err := deps.templates.List(r.Context(), h.fdTenant(r), enabledOnly)
		if err != nil {
			writeError(w, http.StatusInternalServerError, err.Error())
			return
		}
		if tpls == nil {
			tpls = []*freediscovery.ProviderTemplate{}
		}
		writeJSON(w, http.StatusOK, map[string]any{"templates": tpls})

	case http.MethodPost:
		if RequireSuperAdminForWrite(w, r) {
			return
		}
		var req freediscovery.CreateTemplateRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeError(w, http.StatusBadRequest, "invalid JSON body: "+err.Error())
			return
		}
		req.CreatedBy = fdActor(r)
		tpl, err := deps.templates.Create(r.Context(), h.fdTenant(r), &req)
		if err != nil {
			writeError(w, fdStatusFor(err), err.Error())
			return
		}
		writeJSON(w, http.StatusCreated, tpl)

	default:
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
	}
}

// handleFreeDiscoveryTemplateByID GET/PUT/DELETE 单个模板.
func (h *Handler) handleFreeDiscoveryTemplateByID(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid template id")
		return
	}

	switch r.Method {
	case http.MethodGet:
		tpl, err := deps.templates.Get(r.Context(), h.fdTenant(r), id)
		if err != nil {
			writeError(w, fdStatusFor(err), err.Error())
			return
		}
		writeJSON(w, http.StatusOK, tpl)

	case http.MethodPut, http.MethodPatch:
		if RequireSuperAdminForWrite(w, r) {
			return
		}
		var req freediscovery.UpdateTemplateRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeError(w, http.StatusBadRequest, "invalid JSON body: "+err.Error())
			return
		}
		tpl, err := deps.templates.Update(r.Context(), h.fdTenant(r), id, &req)
		if err != nil {
			writeError(w, fdStatusFor(err), err.Error())
			return
		}
		writeJSON(w, http.StatusOK, tpl)

	case http.MethodDelete:
		if RequireSuperAdminForWrite(w, r) {
			return
		}
		if err := deps.templates.Delete(r.Context(), h.fdTenant(r), id); err != nil {
			writeError(w, fdStatusFor(err), err.Error())
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"deleted": true, "id": id})

	default:
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
	}
}

// handleFreeDiscoveryPresets GET 内置供应商预设 (Groq/OpenRouter/...).
func (h *Handler) handleFreeDiscoveryPresets(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	if h.fdDeps(w) == nil {
		return
	}
	type presetView struct {
		ProviderCode string `json:"provider_code"`
		DisplayName  string `json:"display_name"`
		BaseURL      string `json:"base_url"`
		APIType      string `json:"api_type"`
		APIKeyEnv    string `json:"api_key_env"`
		TosVerdict   string `json:"tos_verdict"`
		TosNotes     string `json:"tos_notes"`
	}
	var out []presetView
	for _, code := range freediscovery.ListPresetCodes() {
		p := freediscovery.GetPreset(code)
		if p == nil {
			continue
		}
		out = append(out, presetView{
			ProviderCode: p.ProviderCode,
			DisplayName:  p.DisplayName,
			BaseURL:      p.BaseURL,
			APIType:      string(p.APIType),
			APIKeyEnv:    p.APIKeyEnv,
			TosVerdict:   p.TosVerdict,
			TosNotes:     p.TosNotes,
		})
	}
	writeJSON(w, http.StatusOK, map[string]any{"presets": out})
}

// handleFreeDiscoveryImportOrbi POST 导入 Orbi pi-providers JSON 模板.
// 请求体即 Orbi 模板文件内容: {"providers": {"groq": {...}}}.
func (h *Handler) handleFreeDiscoveryImportOrbi(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodPost {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	if RequireSuperAdminForWrite(w, r) {
		return
	}
	// 限制请求体大小, 防止恶意巨文件 (实际 Orbi 模板通常 < 100 KB).
	r.Body = http.MaxBytesReader(w, r.Body, 1<<20) // 1 MiB
	var file freediscovery.OrbiProviderFile
	if err := json.NewDecoder(r.Body).Decode(&file); err != nil {
		writeError(w, http.StatusBadRequest, "invalid Orbi template JSON: "+err.Error())
		return
	}
	if len(file.Providers) == 0 {
		writeError(w, http.StatusBadRequest, "no providers found in Orbi template")
		return
	}

	tenant := h.fdTenant(r)
	actor := fdActor(r)
	created, failed := 0, 0
	var errs []string
	for code, p := range file.Providers {
		if strings.TrimSpace(code) == "" {
			// 防 panic: 空 provider key 直接跳过 (slice bounds)
			failed++
			errs = append(errs, "<empty provider key>: skipped")
			continue
		}
		// 复用 ValidateCreate 前置校验, 失败记入 errs 而不阻断其它 provider.
		req := &freediscovery.CreateTemplateRequest{
			ProviderCode:   code,
			DisplayName:    freediscovery.TrimmedDisplayName(strings.ToUpper(string([]rune(code)[:1])) + code[1:]),
			BaseURL:        p.BaseURL,
			APIType:        freediscovery.APIType(p.API),
			APIKeyEnv:      "", // Orbi 模板的 apiKey 是 "$VAR" 引用; CreateTemplateRequest 接受后端 env 引用, 由 Create 校验
			ModelsEndpoint: "/models",
			CreatedBy:      actor,
		}
		// Orbi apiKey 字段约定: "$VAR" 表示 env 引用 (保留); 其它值视为字面量, 不作为 env 名解析.
		if strings.HasPrefix(p.APIKey, "$") {
			req.APIKeyEnv = p.APIKey
		}
		if msg := req.ValidateCreate(); msg != "" {
			failed++
			errs = append(errs, code+": "+msg)
			continue
		}
		if _, err := deps.templates.Create(r.Context(), tenant, req); err != nil {
			// 单个 provider 失败不阻断其余导入
			failed++
			errs = append(errs, code+": "+err.Error())
			continue
		}
		created++
	}
	status := "ok"
	if failed > 0 && created > 0 {
		status = "partial" // 部分成功: 调用方可继续查看 errors
	} else if failed > 0 && created == 0 {
		status = "failed"
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"created": created, "failed": failed, "errors": errs, "status": status,
	})
}

// ── 发现任务 ────────────────────────────────────────────────────────────

// handleFreeDiscoveryScan POST 触发一次发现任务.
func (h *Handler) handleFreeDiscoveryScan(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodPost {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	if RequireSuperAdminForWrite(w, r) {
		return
	}
	var body struct {
		TemplateID int64 `json:"template_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body: "+err.Error())
		return
	}
	if body.TemplateID <= 0 {
		writeError(w, http.StatusBadRequest, "template_id is required")
		return
	}

	// 独立超时上下文: 客户端断开不取消扫描 (否则任务永久停留在 running,
	// 且 fail 也无法用已取消的 ctx 写库); 60s 覆盖上游慢响应.
	scanCtx, cancel := context.WithTimeout(context.WithoutCancel(r.Context()), 60*time.Second)
	defer cancel()

	task, err := deps.engine.Run(scanCtx, freediscovery.DiscoveryRequest{
		TemplateID:  body.TemplateID,
		TenantID:    h.fdTenant(r),
		TriggeredBy: fdActor(r),
		TriggerType: freediscovery.TriggerManual,
	})
	if err != nil {
		// 任务已落库 failed 状态时返回 200 + 任务体, 让 UI 能展示失败详情
		if task != nil && task.Status == freediscovery.TaskStatusFailed {
			writeJSON(w, http.StatusOK, task)
			return
		}
		writeError(w, fdStatusFor(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, task)
}

// handleFreeDiscoveryTasks GET 任务列表.
func (h *Handler) handleFreeDiscoveryTasks(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodGet {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	tasks, err := deps.engine.ListTasks(r.Context(), h.fdTenant(r), limit)
	if err != nil {
		// List endpoints have no expected sentinel errors today (failures are DB faults,
		// so fdStatusFor falls through to 500); routed through fdStatusFor for consistency.
		writeError(w, fdStatusFor(err), err.Error())
		return
	}
	if tasks == nil {
		tasks = []*freediscovery.DiscoveryTask{}
	}
	writeJSON(w, http.StatusOK, map[string]any{"tasks": tasks})
}

// handleFreeDiscoveryTask GET 单个任务.
func (h *Handler) handleFreeDiscoveryTask(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodGet {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid task id")
		return
	}
	task, err := deps.engine.GetTask(r.Context(), h.fdTenant(r), id)
	if err != nil {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, task)
}

// handleFreeDiscoveryTaskResults GET 任务发现结果 (?status=pending|all).
func (h *Handler) handleFreeDiscoveryTaskResults(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodGet {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid task id")
		return
	}
	status := r.URL.Query().Get("status")
	if status == "" {
		status = "pending"
	}
	results, err := deps.engine.ListResults(r.Context(), h.fdTenant(r), id, status)
	if err != nil {
		// See handleFreeDiscoveryTasks: no expected sentinels, fdStatusFor falls through to 500.
		writeError(w, fdStatusFor(err), err.Error())
		return
	}
	if results == nil {
		results = []*freediscovery.DiscoveryResult{}
	}
	writeJSON(w, http.StatusOK, map[string]any{"results": results})
}

// handleFreeDiscoveryImport POST 批量导入到 free_resource_catalog.
func (h *Handler) handleFreeDiscoveryImport(w http.ResponseWriter, r *http.Request) {
	deps := h.fdDeps(w)
	if deps == nil {
		return
	}
	if r.Method != http.MethodPost {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return
	}
	if RequireSuperAdminForWrite(w, r) {
		return
	}
	var req freediscovery.ImportRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body: "+err.Error())
		return
	}
	req.TenantID = h.fdTenant(r)
	req.ImportedBy = fdActor(r)
	if req.TaskID <= 0 {
		writeError(w, http.StatusBadRequest, "task_id is required")
		return
	}

	summary, err := deps.importer.Import(r.Context(), req)
	if err != nil {
		writeError(w, fdStatusFor(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, summary)
}

// ── 辅助 ────────────────────────────────────────────────────────────────

// fdActor 操作人标识 (审计; 不含敏感信息).
func fdActor(r *http.Request) string {
	if auth := GetAuthContext(r); auth != nil {
		if auth.Username != "" {
			return auth.Username
		}
		if auth.UserID != 0 {
			return "user-" + strconv.Itoa(auth.UserID)
		}
	}
	return "legacy-admin-key"
}

// fdStatusFor 领域错误 → HTTP 状态码.
//
// 优先用 errors.Is 匹配 sentinel 错误, 避免误把任意含 "not found" 的内部错误
// 映射为 404; 仅当不是 sentinel 时回退到字符串包含判断 (兼容旧错误消息).
func fdStatusFor(err error) int {
	switch {
	case errors.Is(err, freediscovery.ErrTemplateNotFound),
		errors.Is(err, freediscovery.ErrImportTaskNotFound):
		return http.StatusNotFound
	case errors.Is(err, freediscovery.ErrTemplateDisabled),
		errors.Is(err, freediscovery.ErrImportTaskNotReady),
		errors.Is(err, freediscovery.ErrTaskStateConflict):
		return http.StatusConflict
	case strings.Contains(err.Error(), "required"),
		strings.Contains(err.Error(), "invalid"),
		strings.Contains(err.Error(), "must be"),
		strings.Contains(err.Error(), "must start"),
		strings.Contains(err.Error(), "only allows"),
		strings.Contains(err.Error(), "cannot be empty"),
		strings.Contains(err.Error(), "must use"),
		strings.Contains(err.Error(), "userinfo"),
		strings.Contains(err.Error(), "fragment"),
		strings.Contains(err.Error(), "loopback"),
		strings.Contains(err.Error(), "private"),
		strings.Contains(err.Error(), "link-local"),
		strings.Contains(err.Error(), "control characters"),
		strings.Contains(err.Error(), "exceeds maximum length"):
		return http.StatusBadRequest
	default:
		return http.StatusInternalServerError
	}
}
